File: flowersec-swift/Sources/Flowersec/ErrorClassification.swift
public enum RetryAction: String, Equatable, Sendable {
  case retry
  case refreshArtifact = "refresh_artifact"
  case stop
}

public struct ErrorRetryClassification: Equatable, Sendable {
  public let action: RetryAction
  public let callerCanceled: Bool
  public let sessionClosed: Bool

  fileprivate init(
    action: RetryAction,
    callerCanceled: Bool = false,
    sessionClosed: Bool = false
  ) {
    self.action = action
    self.callerCanceled = callerCanceled
    self.sessionClosed = sessionClosed
  }
}

public func classifyConnectError(
  _ error: ConnectError
) -> ErrorRetryClassification {
  switch error {
  case .invalidOptions:
    return ErrorRetryClassification(action: .stop)
  case .canceled:
    return ErrorRetryClassification(action: .stop, callerCanceled: true)
  case .expiredArtifact, .timeout, .connectionFailed:
    return ErrorRetryClassification(action: .refreshArtifact)
  }
}

public func classifySessionError(
  _ error: SessionError
) -> ErrorRetryClassification {
  switch error {
  case .canceled:
    return ErrorRetryClassification(action: .stop, callerCanceled: true)
  case .closed, .goingAway:
    return ErrorRetryClassification(action: .refreshArtifact, sessionClosed: true)
  case .timeout, .resourceExhausted, .streamReset, .rekeyFailed, .livenessFailed:
    return ErrorRetryClassification(action: .retry)
  case .streamRejected, .operationFailed:
    return ErrorRetryClassification(action: .stop)
  }
}

File: flowersec-rust/src/error_classification.rs
use crate::{ConnectError, ConnectErrorCode, SessionError};

/// Carrier-neutral next step for a redacted public failure.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ErrorRetryAction {
    Retry,
    RefreshArtifact,
    Stop,
}

impl ErrorRetryAction {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Retry => "retry",
            Self::RefreshArtifact => "refresh_artifact",
            Self::Stop => "stop",
        }
    }
}

/// Bounded application recovery policy that contains no transport diagnostics.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ErrorRetryClassification {
    pub action: ErrorRetryAction,
    pub caller_canceled: bool,
    pub session_closed: bool,
}

pub const fn classify_connect_error(error: ConnectError) -> ErrorRetryClassification {
    match error.code() {
        ConnectErrorCode::InvalidInput => classification(ErrorRetryAction::Stop, false, false),
        ConnectErrorCode::Canceled => classification(ErrorRetryAction::Stop, true, false),
        ConnectErrorCode::Expired
        | ConnectErrorCode::ResolveFailed
        | ConnectErrorCode::SpendFailed
        | ConnectErrorCode::DialFailed
        | ConnectErrorCode::Timeout
        | ConnectErrorCode::HandshakeFailed => {
            classification(ErrorRetryAction::RefreshArtifact, false, false)
        }
    }
}

pub const fn classify_session_error(error: SessionError) -> ErrorRetryClassification {
    match error {
        SessionError::Canceled => classification(ErrorRetryAction::Stop, true, false),
        SessionError::Closed | SessionError::GoingAway => {
            classification(ErrorRetryAction::RefreshArtifact, false, true)
        }
        SessionError::ResourceExhausted
        | SessionError::StreamReset
        | SessionError::Timeout
        | SessionError::RekeyFailed
        | SessionError::LivenessFailed => classification(ErrorRetryAction::Retry, false, false),
        SessionError::StreamRejected | SessionError::OperationFailed => {
            classification(ErrorRetryAction::Stop, false, false)
        }
    }
}

const fn classification(
    action: ErrorRetryAction,
    caller_canceled: bool,
    session_closed: bool,
) -> ErrorRetryClassification {
    ErrorRetryClassification {
        action,
        caller_canceled,
        session_closed,
    }
}
